aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -0,0 +1,155 @@
+  AT_ASSERT(dense.layout() == kStrided);
+  AT_ASSERT(src.is_sparse_csr());
+  AT_ASSERT(dense.is_cuda());
+
+  TORCH_CHECK(
+      output.is_contiguous(),
+      "out argument must be contiguous, but got: ",
+      output.suggest_memory_format());
+  TORCH_CHECK(
+      output.is_cuda(),
+      "add: expected 'out' to be CUDA tensor, but got tensor on device: ",
+      output.device());
+
+  TORCH_CHECK(
+      src.is_cuda(),
+      "add: expected 'other' to be a CUDA tensor, but got tensor on device: ",
+      src.device());
+
+  TORCH_CHECK(
+      dense.sizes().equals(src.sizes()),
+      "add: expected 'self' and 'other' to have same size, but self has size ",
+      dense.sizes(),
+      " while other has size ",
+      src.sizes(),
+      " (FYI: op2-sparse addition does not currently support broadcasting)");
+
+  auto commonDtype = promoteTypes(dense.scalar_type(), src.scalar_type());
+  TORCH_CHECK(
+      canCast(commonDtype, output.scalar_type()),
+      "Can't convert result type ",
+      commonDtype,
+      " to output ",
+      output.scalar_type(),
+      " in add operation");
+
+  Tensor src_values = src.values().to(commonDtype);
+  Tensor src_crow_indices = src.crow_indices();
+  Tensor src_col_indices = src.col_indices();
+
+  output.resize_as_(dense);
+  Tensor resultBuffer = output;
+  Tensor valuesBuffer = src_values.to(commonDtype);
+
+  if (output.scalar_type() != commonDtype) {
+    resultBuffer = dense.to(commonDtype);
+  } else if (!is_same_tensor(output, dense)) {
+    resultBuffer.copy_(dense);
+  }
+  AT_DISPATCH_ALL_TYPES(
+      commonDtype,
+      "add_out_op2_sparse_csr",
+      [&src_values, &output, &alpha, &src_crow_indices, &src_col_indices]() {
+        AT_DISPATCH_INDEX_TYPES(
+            src_crow_indices.scalar_type(),
+            "csr_add_out_crow_indices",
+              [&src_values, &output, &alpha, &src_crow_indices, &src_col_indices]() {
+                scalar_t* values_accessor = src_values.data_ptr<scalar_t>();
+                scalar_t* out_ptr = output.data_ptr<scalar_t>();
+                scalar_t cast_value = alpha.to<scalar_t>();
+
+                index_t* crow_indices_accessor = src_crow_indices.data_ptr<index_t>();
+                index_t* col_indices_accessor = src_col_indices.data_ptr<index_t>();
+                int64_t out_storage_offset = output.storage_offset();
+
+                auto out_strides = output.strides();
+                int64_t out_strides0 = out_strides[0];
+                int64_t out_strides1 = out_strides[1];
+
+                cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+                auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+                auto policy = thrust::cuda::par(allocator).on(stream);
+
+               thrust::for_each(
+                    policy,
+                    thrust::make_counting_iterator(int64_t(0)),
+                    thrust::make_counting_iterator(int64_t(src_crow_indices.size(0) - 1)),
```

Comment:
NB: as is the classic problem with parallelizing sparse algorithms, this could be wildly imbalanced if the sparsity pattern varies a lot between rows. There isn't an easy fix for this, but it would be nice for the problem to be remarked upon in code.